File: tests/xcalablemp/global-view/shadow-reflect/F/reduce_shadow.f90
program test

  !$xmp nodes p(4,4)

  integer m(4)
  integer n
  
  !$xmp template t(:,:)
  !$xmp distribute t(gblock(*),gblock(*)) onto p

  integer, allocatable :: a(:,:)
  !$xmp align a(i,j) with t(i,j)
  !$xmp shadow a(1:1,1:1)

  integer, allocatable :: b(:,:)

  integer :: result = 0

  n = 32
  m = (/4,4,8,16/)
  
  !$xmp template_fix(gblock(m),gblock(m)) t(n,n)
  allocate (a(n,n))

  allocate (b(n,n))
  
  !$xmp array on t(1:n,1:n)
  a(1:n,1:n) = 1

  do j = 1, n
     do i = 1, n
        b(i,j) = 1
        if (i == 1 .OR. i == 4 .OR. &
            i == 5 .OR. i == 8 .OR. &
            i == 9 .OR. i == 16 .OR. &
            i == 17 .OR. i == 32) then
           b(i,j) = b(i,j) + 1
        end if
        if (j == 1 .OR. j == 4 .OR. &
            j == 5 .OR. j == 8 .OR. &
            j == 9 .OR. j == 16 .OR. &
            j == 17 .OR. j == 32) then
           b(i,j) = b(i,j) + 1
        end if
     end do
  end do
  
  !$xmp reflect (a) width(/periodic/1,/periodic/1)

  !$xmp reduce_shadow (a) width(/periodic/1,/periodic/1) async(10)

  !$xmp wait_async (10)

  !$xmp loop (i,j) on t(i,j) reduction(+:result)
  do j = 1, n
     do i = 1, n
        if (a(i,j) /= b(i,j)) then
           result = 1
        end if
     end do
  end do

  !$xmp task on p(1,1)
  if (result /= 0) then
     write(*,*) "ERROR"
     call exit(1)
  else
     write(*,*) "PASS"
  endif
  !$xmp end task

end program test

File: tests/XMP/global-view/shadow-reflect/C-Round/reduce_shadow.c
#include <stdio.h>
#include <stdlib.h>
#include "xmp.h"

int main(){

#pragma xmp nodes p(4,4)

  int m[4];
  int n;
  
#pragma xmp template t(:,:)
#pragma xmp distribute t(gblock(*),gblock(*)) onto p

  n = 32;
  m[0] = 4; m[1] = 4; m[2] = 8; m[3] = 16;

  int (*a)[n];
#pragma xmp align a[i][j] with t(i,j)
#pragma xmp shadow a[1:1][1:1]

  int (*b)[n];
  
#pragma xmp template_fix(gblock(m),gblock(m)) t(0:n-1,0:n-1)
  a = (int (*)[n])xmp_malloc(xmp_desc_of(a), n, n);

  b = (int (*)[n])malloc(sizeof(int)*n*n);
  
#pragma xmp array on t(0:n-1,0:n-1)
  a[0:n][0:n] = 1;

  for (int i = 0; i < n; i++){
    for (int j = 0; j < n; j++){
      b[i][j] = 1;
      if (i == 0 || i == 3 || 
	  i == 4 || i == 7 || 
	  i == 8 || i == 15 || 
	  i == 16 || i == 31){
	b[i][j]++;
      }
      if (j == 0 || j == 3 || 
	  j == 4 || j == 7 || 
	  j == 8 || j == 15 || 
	  j == 16 || j == 31){
	b[i][j]++;
      }
    }
  }

#pragma xmp reflect (a) width(/periodic/1, /periodic/1)

#pragma xmp reduce_shadow (a) width(/periodic/1,/periodic/1) async(10)

#pragma xmp wait_async (10)

  int result = 0;
  
#pragma xmp loop (i,j) on t(i,j) reduction(+:result)
  for (int i = 0; i < n; i++){
    for (int j = 0; j < n; j++){
      if (a[i][j] != b[i][j]) result = 1;
    }
  }

#pragma xmp task on p(1,1)
  {
    if (result == 0){
      printf("PASS\n");
    }
    else{
      fprintf(stderr, "ERROR\n");
      exit(1);
    }
  }

  return 0;
  
}
